scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,256 @@
+use crate::cdc_types::{GenerationTimestamp, StreamID};
+use anyhow;
+use async_trait::async_trait;
+use futures::future::RemoteHandle;
+use futures::FutureExt;
+use scylla::frame::response::result::CqlValue::Timestamp;
+use scylla::prepared_statement::PreparedStatement;
+use scylla::Session;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::sleep;
+
+/// Alias representing timestamp.
+/// (Last read timestamp, one of stream ids tracked by stream reader, generation of stream)
+pub type Checkpoint = (Duration, StreamID, GenerationTimestamp);
+
+/// Starts creating checkpoints starting from now.
+/// Spawns a new task that periodically creates new checkpoint.
+/// Receives value of the next time window via channel.
+/// We are using tokio::sync::watch channel, because we only need the latest value.
+/// Returns handle to check on this newly created task.
+pub fn start_making_checkpoints(
+    checkpoint_maker: Arc<dyn CDCCheckpointMaker>,
+    receiver: tokio::sync::watch::Receiver<Checkpoint>,
+    saving_period: Duration,
+) -> RemoteHandle<anyhow::Result<()>> {
+    let (fut, handle) = async move {
+        loop {
+            let mut checkpoint = receiver.borrow().clone();
+
+            // We are given timestamp of the beginning of the next read window,
+            // so we have to create a checkpoint with lower timestamp.
+            if checkpoint.0 > Duration::from_millis(1) {
+                checkpoint.0 -= Duration::from_millis(1);
+                checkpoint_maker.make_checkpoint(checkpoint).await?;
+            }
```

Comment:
Despite the comment, I still don't understand why this is needed. Could you explain why we need to subtract one millisecond? Why can't we just store the timestamp just at it is?